File: EmotionEngine/src/main/java/co/q64/emotion/lang/Instruction.java
package co.q64.emotion.lang;

import java.util.function.Consumer;

import com.google.auto.factory.AutoFactory;
import com.google.auto.factory.Provided;

import co.q64.emotion.lang.opcode.Opcodes;
import co.q64.emotion.lang.value.Value;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

import javax.inject.Inject;

@Accessors(fluent = true)
public class Instruction {
    protected @Inject Opcodes opcodes;

    private @Getter @Setter Integer opcode;
    private @Setter Value value;
    private Consumer<Stack> executor;

    protected @Inject Instruction() {}

    public void execute(Stack stack) {
        if (value != null) {
            stack.push(value);
            return;
        }
        if (opcode != null) {
            if (executor == null) {
                this.executor = opcodes.getExecutor(opcode);
            }
        }
        if (executor != null) {
            executor.accept(stack);
        }
        // no-op
    }

    public String getInstruction() {
        if (value != null) {
            return "load " + value.toString();
        }
        if (executor != null) {
            return opcodes.getName(opcode).orElse("undefined");
        }
        return "nop";
    }

    @Override
    public String toString() {
        return getInstruction();
    }
}

File: emotion-core/src/main/kotlin/co/q64/emotion/core/lang/Instruction.kt
package co.q64.emotion.core.lang

import co.q64.emotion.core.opcode.OpcodeMarker
import co.q64.emotion.core.value.Value

sealed interface Instruction {
    val marker: OpcodeMarker? get() = null
    fun execute(program: Program)

    data class Execute(
        private val pending: PendingInstruction
    ) : Instruction {
        override val marker: OpcodeMarker? get() = pending.marker

        override fun execute(program: Program) = with(pending.resolve(program)) {
            program.closure()
        }

        override fun toString(): String = "$pending"
    }

    data class Load(
        private val value: Value
    ) : Instruction {
        override fun execute(program: Program) {
            program.push(value)
        }

        override fun toString(): String = "load $value"
    }

    object NoOp : Instruction {
        override fun execute(program: Program) = Unit
        override fun toString(): String = "no-op"
    }
}